test/AutoDiff/SIMD.swift

```diff
@@ -0,0 +1,194 @@
+// RUN: %target-run-simple-swift
+// REQUIRES: executable_test
+
+import StdlibUnittest
+#if os(macOS)
+import Darwin.C
+#else
+import Glibc
+#endif
+
+var SIMDTests = TestSuite("SIMD")
```

Comment:
A testing idea that needn't block progress:

Currently, all tests seem to operate on `SIMD<Float>`. It would be nice to test differentiation of generic functions over `SIMD`, with the appropriate generic constraints. Example:
```swift
func testInit<Scalar, SIMDType>(x: Scalar) -> SIMDType
  where SIMDType : Differentiable,
        SIMDType.TangentVector : SIMD,
        SIMDType.Scalar == Scalar,
        Scalar : BinaryFloatingPoint & Differentiable,
        Scalar.TangentVector : BinaryFloatingPoint,
        SIMDType.TangentVector == SIMDType {
  return SIMDType.init(repeating: x)
}
```

Defining generic functions also allows us to test multiple `SIMD` and `Scalar` type combinations, e.g. `SIMD2<Double>`.

Reply:
Good idea! I can definitely a couple tests regarding generics. It would be especially good to do it given the way `SIMD` is implemented with a protocol and then structs.

Reply:
@dan-zheng the constraints on the vjpInit are the following:
```swift
extension SIMD
  where Self : Differentiable,
        Self.TangentVector : SIMD,
        Scalar : BinaryFloatingPoint & Differentiable,
        Self.TangentVector == Self,
        Scalar.TangentVector == Scalar {
  @usableFromInline
  static func _vjpInit(repeating value: Scalar)
    -> (Self, (TangentVector) -> Scalar.TangentVector) {
      return (Self(repeating: value), { v in v.sum() })
  }
}
```
Thus I think a generic init function would look like this:
```swift
func testInit<Scalar, SIMDType>(x: Scalar) -> SIMDType
  where SIMDType : Differentiable,
        SIMDType.TangentVector : SIMD,
        Scalar : BinaryFloatingPoint & Differentiable,
        SIMDType.TangentVector == SIMDType,
        Scalar.TangentVector == Scalar {
  return SIMDType.init(repeating: x)
}
```
However I'm getting an error saying it's expecting me to use the `init(from decoder: Decoder)` constructor:
```
simd.swift:7:27: error: incorrect argument label in call (have 'repeating:', expected 'from:')
      return SIMDType.init(repeating: x)
                          ^~~~~~~~~~
                           from
```
Is there something wrong in my constraints, or is this a bug?

Reply:
@dan-zheng quick ping when you have a minute

Reply:
The `SIMDType.Scalar == Scalar` constraint is missing.
(The original program I shared does have that constraint.)

Reply:
Either there is a bug or there is another constraint missing (looking right now), but I can't mark the function you posted as differentiable:
```swift
@differentiable
func testInit<Scalar, SIMDType>(x: Scalar) -> SIMDType
  where SIMDType : Differentiable,
        SIMDType.TangentVector : SIMD,
        SIMDType.Scalar == Scalar,
        Scalar : BinaryFloatingPoint & Differentiable,
        Scalar.TangentVector : BinaryFloatingPoint,
        SIMDType.TangentVector == SIMDType {
  return SIMDType.init(repeating: x)
}
```

```
simd.swift:24:19: error: expression is not differentiable
  return SIMDType.init(repeating: x)
                  ^
simd.swift:24:19: note: function call is not differentiable because generic requirements are not met
  return SIMDType.init(repeating: x)
                  ^
```

Reply:
That's why my function had different constraints, as I was running into this issue and was messing around with the constraints.

Reply:
Added generic tests in [`d73f3da`](https://github.com/apple/swift/pull/24786/commits/d73f3dab585cb17ea128e215a634bf2c8d6cf985).